File: scala/variable-length-quantity/src/main/scala/VariableLengthQuantity.scala
import scala.annotation.tailrec

object VariableLengthQuantity {
  private val SevenBitsMask = 0x7f
  private val EightBitMask = 0x80

  def encode(bytes: List[Int]): List[Int] = bytes.flatMap(encodeHelper)

  private def encodeHelper(byte: Int): List[Int] = {
    @tailrec
    def encodeByte(bytes: List[Int], remainder: Int): List[Int] = {
      if (remainder == 0) {
        bytes
      } else {
        val maskedBytes =
          if (bytes.isEmpty)
            remainder & SevenBitsMask
          else
            remainder & SevenBitsMask | EightBitMask

        encodeByte(maskedBytes :: bytes, remainder >>> 7)
      }
    }

    if (byte == 0) List(0) else encodeByte(List(), byte)
  }

  def decode(encodedBytes: List[Int]): Either[String, List[Int]] = decodeHelper(0x0, List(), encodedBytes)

  @tailrec
  private def decodeHelper(decodedValue: Int, decodedBytes: List[Int], encodedBytes: List[Int]): Either[String, List[Int]] = {
    encodedBytes match {
      case Nil =>
        if (decodedValue == 0 && decodedBytes.nonEmpty)
          Right(decodedBytes.reverse)
        else
          Left("Incomplete byte sequence")
      case x :: xs =>
        if ((decodedValue & 0xfe000000) > 0) {
          Left("Overflow exception")
        }
        else {
          val value = (decodedValue << 7) | (x & SevenBitsMask)
          if ((EightBitMask & x) == 0)
            decodeHelper(0, value :: decodedBytes, xs)
          else
            decodeHelper(value, decodedBytes, xs)
        }
    }
  }
}

File: typescript/variable-length-quantity/variable-length-quantity.ts
export default class VLQ {
  private static readonly SevenBitsMask = 0x7f;
  private static readonly EightBitMask = 0x80;

  public static encode(integers: number[]): number[] {
    const encodedBytes: number[] = [];
    integers.forEach(integer =>
      encodedBytes.push(...VLQ.encodeInteger(integer))
    );

    return encodedBytes;
  }

  private static encodeInteger(integer: number): number[] {
    if (integer == 0) {
      return [0];
    }

    let bytes = [];

    while (integer > 0) {
      var tmp = integer & VLQ.SevenBitsMask;
      integer >>>= 7;

      if (bytes.length > 0) {
        tmp |= VLQ.EightBitMask;
      }

      bytes.push(tmp);
    }

    bytes.reverse();
    return bytes;
  }

  public static decode(encodedBytes: number[]): number[] {
    const decodedIntegers: number[] = [];
    let tmp = 0;
    for (let index = 0; index < encodedBytes.length; index++) {
      if ((tmp & 0xfe000000) > 0) {
        throw new Error("Overflow exception.");
      }
      const b = encodedBytes[index];
      tmp = (tmp << 7) | (b & 0x7f);
      if ((b & 0x80) == 0) {
        decodedIntegers.push(tmp >>> 0);
        tmp = 0;
      } else if (index == encodedBytes.length - 1) {
        throw new Error("Incomplete sequence.");
      }
    }
    return decodedIntegers;
  }
}
